### ISSUE TITLE
Training on NCHW and testing on NHWC

### ISSUE THREAD

**Issue by tumusudheer** [xiaodiu2010/TextBoxes-TensorFlow#21]
Hi @xiaodiu2010 ,

I've trained my network using NCHW format on GPUs. I did freeze the graph and tested on GPU and getting good results. But I'm getting an error when I want to use the same graph to test on CPU, may be because CPU doesn't support NCHW. Is it possible to convert my graph trained on GPU so that it works (to test) on CPU also ?

I also retrained using NHWC format using GPUs ( by specifying data_format argument = 'NHWC' on Textbox_train.py and modifying load_batch.py) and tested on CPU using this graph. but I'm not getting good results with the graph I trained on CPU. Are there any bug in the code path of NHWC data_format ?

Thank you.

**Comment by tumusudheer** [comment]
Not sure what I'm doing wrong because my trained graph with NHWC is not working as the one trained using NCHW.  Do I need to assign any other flags apart from data_format = 'NHWC' in Textbox_train.py

I'm finetuning from vgg_16 checkpoint when training for CPU with NHWC format. Do I need to find a trained vggnet in NHWC format when training using NHWC format ? Or can I use the same vgg_16 pre-trained net when training with NHWC format and NCHW format ?

**Comment by xiaodiu2010** [comment]
The pre-train model is irrelative to format. Can u give me more details. Are there any bugs? error info？

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010,

I've trained the model using GPUS with NSHW format and when I tested the model, I'm getting good results using the test_image.py scipt
[test_image.py.txt](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/files/1312705/test_image.py.txt)

When I want to test the same code on a machine that has only CPU, I'm getting error that NCHW operations are not permitted on CPUs.

There are the steps I've followed to test:

1. Frozen the graph using the net create as follows:

```
  with tf.Graph().as_default() as graph:
    network_fn = nets_factory.get_network('text_box_300')

    placeholder = tf.placeholder(name='input', dtype=tf.float32,
                                 shape=[ None,
                                        None, 3])
    net = network_fn()
    out_shape = (300,300)
    image_pre, labels_pre, bboxes_pre, bbox_img, num = txt_preprocessing.preprocess_for_eval(placeholder, None, None, None, None, out_shape, resize=txt_preprocessing.Resize.WARP_RESIZE,use_whiten=True)
    image_4d = tf.expand_dims(image_pre, 0)
    image_4d = tf.transpose(image_4d, perm=(0, 3, 1, 2))

    #net.net(image_4d, is_training=False, reuse=None,use_batch=True)
    with slim.arg_scope(net.arg_scope(data_format='NCHW')):
        localisations, predictions,  _ = net.net(image_4d, is_training=False, reuse=None,use_batch=True)
        graph_def = graph.as_graph_def()
        with gfile.GFile('binary_cpu.pb', 'wb') as f:
            f.write(graph_def.SerializeToString())

```

2. restore the weights from trained checkpoint
3. Running the forward pass on a CPU machine.

 The precise error is as follows:
```
2017-09-18 17:57:48.890761: E tensorflow/core/common_runtime/executor.cc:644] Executor failed to create kernel. Invalid argument: Default MaxPoolingOp only supports NHWC.
         [[Node: text_box_300/pool1/MaxPool = MaxPool[T=DT_FLOAT, data_format="NCHW", ksize=[1, 1, 2, 2], padding="SAME", strides=[1, 1, 2, 2], _device="/job:localhost/replica:0/task:0/cpu:0"](text_box_300/conv1/conv1_2/Relu)]]
2017-09-18 17:57:48.892768: E main_textd_test.cc:457] Running model failed: Invalid argument: Default MaxPoolingOp only supports NHWC.
         [[Node: text_box_300/pool1/MaxPool = MaxPool[T=DT_FLOAT, data_format="NCHW", ksize=[1, 1, 2, 2], padding="SAME", strides=[1, 1, 2, 2], _device="/job:localhost/replica:0/task:0/cpu:0"](text_box_300/conv1/conv1_2/Relu)]]

```

Because of the above error, I thought I need to train using NHWC format to test on CPUs. I started training on a CPU machine, But I'm not getting good results when I tested with the graph I trained on CPU. Is it possible to use the trained model I trained on GPUs to test on CPU machine ?

```
    with slim.arg_scope(net.arg_scope(data_format='NCHW')):
        localisations, predictions,  _ = net.net(image_4d, is_training=False, reuse=None,use_batch=True)
```
The above line creates a net such that the maxPool layer takes NCHW format as input. When you run the forward pass on the net with NHWC as input, I guess it is giving error.

Thank you,

**Comment by xiaodiu2010** [comment]
Sure, You can train your model on GPU using NHMC format. And this is can be tested on cpus.

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010,

I did train the model using NHWC format on GPUS twice. The results are not at all looking good. The Model is having loss ~ 4 even at 220k iterations. So I stopped and tested with 220k model twice. Both times the results not even comparable to the model I got trained with NCHW format. Any thoughts on why you think I'm not getting good results when I trained on NHWC format on GPUs ? Attaching the results I got with both NCHW format and NHWC format. You can see the results with NHWC trained model are very bad.
![results_nchw](https://user-images.githubusercontent.com/3826735/30571754-9e731aaa-9c9e-11e7-826b-ae2aa61e6f6a.jpg)
![results_nhwc](https://user-images.githubusercontent.com/3826735/30571756-a2bb9a6a-9c9e-11e7-989a-4c7af43f2eb5.jpg)



Thank you.

**Comment by xiaodiu2010** [comment]
So interesting. Did you use batch norm?

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

Yeah I'm using batch_norm. Here is the command I'm using to run the training:

```
CHECKPOINT_PATH=./checkpoints/vgg_16.ckpt
DATASET_DIR=./data/sythtext/
TRAIN_DIR=./logs/train
TF_ENABLE_WINOGRAD_NONFUSED=1 CUDA_VISIBLE_DEVICES=0 setsid python Textbox_train.py \
        --train_dir=${TRAIN_DIR} \
        --dataset_dir=${DATASET_DIR} \
        --save_summaries_secs=60 \
        --save_interval_secs=1800 \
        --num_readers=8 \
        --num_preprocessing_threads=8 \
        --optimizer=momentum \
        --weight_decay=0.0005 \
        --learning_rate=0.001 \
        --batch_size=32 \
        --num_samples=800000 \
        --gpu_memory_fraction=0.80 \
        --max_number_of_steps=500000 \
        --use_batch=True \
        --num_clones=1 \
        --negative_ratio=3 \
        --fine_tune=True \
        --checkpoint_path=${CHECKPOINT_PATH} \
        --checkpoint_model_scope=vgg_16 \
        --ignore_missing_vars=True \
```

I'm using the vgg16 downloaded from here: [http://download.tensorflow.org/models/vgg_16_2016_08_28.tar.gz](http://download.tensorflow.org/models/vgg_16_2016_08_28.tar.gz). Do I need to finetine from a vgg checkpoint trained model that has been trained with NHWC format ?

Also, in your previous reply you have mentioned the pre-train model is irrelative to format. Does it mean that I can use the model that I trained using NCHW format to test the data in NHWC format on CPU ? may I know how can I do that ?

**Comment by xiaodiu2010** [comment]
Sorry for that. This is a bug.  Try this, change the following line in load_batch.py
image, glabels, gbboxes,num = 
txt_preprocessing.preprocess_image(image, glabels, gbboxes, height, width, out_shape,use_whiten=FLAGS.use_whiten,is_training=is_training)
into
image, glabels, gbboxes,num = 
txt_preprocessing.preprocess_image(image, glabels, gbboxes, height, width, out_shape,use_whiten=FLAGS.use_whiten,is_training=is_training, data_format=FLAGS.data_format)

I forgot including format when loading batch

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

I've already made that change in load_batch.py
```
image, glabels, gbboxes,num = 
txt_preprocessing.preprocess_image(image,  glabels, gbboxes, height, width,                                                                                   out_shape,use_whiten=FLAGS.use_whiten,is_training=is_training,data_format=FLAGS.data_format)
```

was the code I used for my training.  This is the only change I've made. Do I need to change any other part of the code or do modification ?

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

I started debugging the problem and I found one issue here:[https://github.com/tumusudheer/TextBoxes-TensorFlow/blob/master/nets/txtbox_300.py#L288]( https://github.com/tumusudheer/TextBoxes-TensorFlow/blob/master/nets/txtbox_300.py#L288)

When you invoke batch_norm, the data_format param is not being passed, so this condition is becoming true,  always data_format is being passed as NHWC no matter what the input data_format is [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L312](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L312) 

Thank you...

**Comment by xiaodiu2010** [comment]
@tumusudheer I don't think this is the issue, the data format is included in arg_scope method. And this method includes slim.conv2d already.  This issue occurs someplace I try to do some transpose or other operation which only supports HCWH, I think. Have you tried to train with NHWC format without batch norm ?

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

I just started debugging why the model with NHWC was not working for me as it was with NCHW. Just wanted to let you know about this issue as the params_shape is being defined by the data_format here:
[https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L315](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L315)

and also the params_shape is going to be the shape of other batch_norm parms like moving_mean here: [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L366](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/layers/python/layers/layers.py#L366)

 This may not be the reason as you were saying. I'll keep debugging and keep posting here. If you find the reason, please let me now.



 Also, it is a good idea to train with NHWC format without batch_norm, I'll try and start training tonight. But wouldn't the accuracy go down as you said in your notes that batch_norm would improve accuracy by 5-10% ?

**Comment by xiaodiu2010** [comment]
@tumusudheer  try add `slim.batch_norm` in ssd_arg_scope method,
` with slim.arg_scope([custom_layers.pad2d,
                                 custom_layers.l2_normalization,
                                 custom_layers.channel_to_last,
                                 slim.batch_norm],
                                data_format=data_format) as sc:`
and how about your result without batch_norm?

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010,

I've changed the ssd_arg_scope method as per your suggestion. I couldn't run training without batch_norm with NHWC format over the weekend. I just started training without batch_norm.
Here is the command I've used:
```
CHECKPOINT_PATH=./checkpoints/vgg_16.ckpt
DATASET_DIR=./data/train_2.0/
TRAIN_DIR=./logs/train_no_batch_norm/
TF_ENABLE_WINOGRAD_NONFUSED=1 CUDA_VISIBLE_DEVICES=0 setsid python Textbox_train.py \
        --train_dir=${TRAIN_DIR} \
        --dataset_dir=${DATASET_DIR} \
        --save_summaries_secs=60 \
        --save_interval_secs=1800 \
        --num_readers=8 \
        --num_preprocessing_threads=8 \
        --optimizer=momentum \
        --weight_decay=0.0005 \
        --learning_rate=0.001 \
        --batch_size=32 \
        --num_samples=800000 \
        --gpu_memory_fraction=0.80 \
        --max_number_of_steps=500000 \
        --use_batch=False \
        --num_clones=1 \
        --negative_ratio=3 \
        --fine_tune=True \
        --checkpoint_path=${CHECKPOINT_PATH} \
        --checkpoint_model_scope=vgg_16 \
        --ignore_missing_vars=True \

```
let me know if there is anything wrong in the command to train without batch_norm. Of course, I've changed the data_format to NHWC in Textbox_train.py

Also, I started training with batch_norm and slim.batch_norm included in the arg_scope (as per your suggestion) with NCHW and also NHWC formats. So in total I've 3 different kinds of training going on. I'll update you about the accuracy when training is

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

I was able to figureout to restore a checkpoint that has trained on NCHW with the model that takes NHWC as input. Now I can train using NCHW on GPUs and can run the graph on CPU with NHWC. The batch_norm not being added to the arg_scope is the issue. But when I added batch_norm to the arg_scope, the training is converging very slowly, not sue why though.

A couple of quick questions:
1. I would like the authors did in the original paper, use both 300 and 700 nets and finally use NMS to aggregate boxes. How can I do this kind of multi scaling using both 300 and 700 nets ? Do I need to create two different nets in Textboxes_train.py and create two clones  like here: [https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/Textbox_train.py#L291](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/Textbox_train.py#L291). It would be great if you can explain briefly how to make use both 300 and 700 nets to detect boxes

2. How to determine the gradient_multiplier (values) that should be in multiplier_700.pkl ? Earlier, you have mentioned to follow [https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/nets/multiplier_300.pkl](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/nets/multiplier_300.pkl) file to create multiplier_700.pkl file. Can you let me know how to determine the values/gradient multipliers that should be in multiplier_700.pkl

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

can you please let me know what should be the gradient weights for tensors in multiplier_700.pkl file ? I can get the name of tensors and use your multiplier_300.pkl as reference, but may I know how to calculate the values for gradient multipliers in multiplier_700.pkl ?

Thanks in advance !!!

**Comment by xiaodiu2010** [comment]
@tumusudheer 
About the multi-scales, as mentioned in the paper, you can just train your model on 300 scale level and use it to inference on 700 scale.  Cause ssd is a fully convolutional network. Also you may try to fine tune your 300 model on 700 level by frozen all layers except textbox layer.

About the gradient_multiplier values,  I used the original ones in ssd prototxt file in [here](https://github.com/MhLiao/TextBoxes/blob/TextBoxes/examples/TextBoxes/deploy.prototxt)

**Comment by tumusudheer** [comment]
Hi  @xiaodiu2010 ,

Thank you very much. I'll go through the original ssd prototxt you linked to get gradient_multiplier values. To fine tune my 300 model with 700 level, These are the following steps I'm going to follow, please let me know if I'm doing it in a right way:
1. Get list of all tensors being prepared in [https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/nets/txtbox_300.py](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/nets/txtbox_300.py)

2. Use the list of tensor names (comma separated) as value for `trainable_scopes`  parameter: [https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/Textbox_train.py#L188](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/Textbox_train.py#L188)

3. Pass text_box_700 for `model_name`, my trained 300 check point path `checkpoint_path`  and ignore `checkpoint_model_scope`

Please correct me if I'm doing any thing wrong. Thank you very much once again.

Thank you

**Comment by xiaodiu2010** [comment]
@tumusudheer  
you can list all trainable_scopes. But it will be a long list. instead, you can just assign 0 learning multiplier on those variables you don't want to train again. So those variables won't be updated. 
`checkpoint_model_scope` should equal to the scope name in checkpoint you used. In this case, it is `text_box_300`

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

Thank you very much. You mean set gradient_multipliers for the tensors which  we do not train to zero [here](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/Textbox_train.py#L331)  ? May I know How do I mention learning_rate to be zero for variables we do not want to train. I found the following two ways, do you recommend any one of the following:

1. Use tf.variable(trainable=False) like [here](https://www.tensorflow.org/programmers_guide/variables#variable_collections)
2. Use tf.stop_gradient like [here](https://www.tensorflow.org/api_docs/python/tf/stop_gradient)

Do you recommend any of the above or any other way you know to set learning_rate to zero for variables you do not want to train ?

Thank you.

**Comment by xiaodiu2010** [comment]
You don't need to change graph.Just set the multiplier to 0 for some variables, then those variables won't be trained.  learning_rate * 0 = 0(no update)

**Comment by tumusudheer** [comment]
Hi @xiaodiu2010 ,

Thank you very much for your reply and Sorry for my late reply. When you mentioned to fine tune 300 model on  700 level, you have asked me to pass text_box_700 value for parameter [here](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/Textbox_train.py#L160) which will eventually make the input images resized to 700 X 700 scale. After the training I'll have a 700 model trained for 700 scale input images and I already have 300 model that has been trained on 300 images. i.e. I'll have two model trained for two different sizes. Instead of this, is it possible to have one single model (eg: text_box_300 net) trained for different image sizes so that I can use that one single model for inference/testing instead of having multiple models to run inference on multiple image sizes ?

The text_box_300 net has image_shape 300X300 [here](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/nets/txtbox_300.py#L64) and text_box_700 net has image_shape 700X700 [here](https://github.com/xiaodiu2010/TextBoxes-TensorFlow/blob/master/nets/txtbox_700.py#L63) and also the variables/tensors in each net have different names because of different scopes. So I guess I can not pass image_shape has parameter for text_box_300 net to train it for different image shapes/sizes ?

I really thank you very much for this model and also answering all my questions. Thank you very much.

**Comment by hellotf** [comment]
@tumusudheer Have you known how to get multiplier_700.pkl already? I've trained 300*300 net on my own dataset, however the result is not good enough.
